### PR TITLE
Compatibility for clang

### PR DESCRIPTION
**Pull request by ikanago**
To solve #431, adapt clang to build this project for arm64 on x86 via GitHub Actions.
I created this PR based on the same branch with #224 because we cannot reopen #224.

### CODE REVIEW

**Comment by forest1040** [comment]
Thank you! but my enviroment was ocured compile error.
`./script/build_gcc.sh` is success.
`pip install .` is error.

I attache error log.

```
      /Users/tmori/work/qulacs/src/csim/type.hpp:19:22: error: expected namespace name
[pip_install.log](https://github.com/qulacs/qulacs/files/10241914/pip_install.log)

      using namespace std::complex_literals;

      In file included from /Users/tmori/work/qulacs/src/cppsim/gate_factory.hpp:12:
      /Users/tmori/work/qulacs/src/cppsim/gate_general.hpp:62:16: error: 'auto' not allowed in lambda parameter
                  [](auto gate) { return gate->copy(); });
                     ^~~~
      /Users/tmori/work/qulacs/src/cppsim/gate_general.hpp:229:16: error: 'auto' not allowed in lambda parameter
                  [](auto gate) { return gate->copy(); });
                     ^~~~
      /Users/tmori/work/qulacs/src/cppsim/gate_general.hpp:370:16: error: 'auto' not allowed in lambda parameter
                  [](auto gate) { return gate->copy(); });
                     ^~~~
```

It's my compiler envilonment.
```
      -- CMAKE_CXX_COMPILER_ID = AppleClang
      -- CMAKE_C_COMPILER_VERSION = 14.0.0.14000029
      -- CMAKE_CXX_COMPILER_VERSION = 14.0.0.14000029
      -- CMAKE_CXX_COMPILER_ID = AppleClang
      -- CMAKE_CXX_COMPILER_ID = AppleClang
      -- CMAKE_HOST_SYSTEM_PROCESSOR = arm64
```

**Comment by forest1040** [comment]
[pip_install.log](https://github.com/qulacs/qulacs/files/10241923/pip_install.log)

**Comment by forest1040** [comment]
@ikanago @templepmet 
https://github.com/qulacs/qulacs/blob/bugfix/update_cmakelist_for_clang/test/gpusim/test_update_pauli.cpp
It's merge conflict by gpusim.
Please check it.

File: test/gpusim/test_update_pauli.cpp [hunk 1]
```diff
@@ -44,8 +44,14 @@ TEST(UpdateTest, SingleQubitPauliRotationTest) {
     const ITYPE dim = 1ULL << n;
     const UINT max_repeat = 10;
 
+<<<<<<< HEAD
```
Comment (KowerKoint) [inline test/gpusim/test_update_pauli.cpp:47]: Unresolved conflict in this file.

**Comment by ikanago** [comment]
@forest1040 what is the error you are facing?
I have no error with `-std=c++14` on this branch.

File: script/build_clang.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+
+GCC_COMMAND=clang
+GXX_COMMAND=clang++
+
+mkdir ./build
+cd ./build
+cmake -G "Unix Makefiles" -D CMAKE_C_COMPILER=$GCC_COMMAND -D CMAKE_CXX_COMPILER=$GXX_COMMAND -D CMAKE_BUILD_TYPE=Release -D USE_GPU:STR=No USE_OMP:STR=No ..
```
Comment (KowerKoint) [inline script/build_clang.sh:8]:
```suggestion
cmake -G "Unix Makefiles" -D CMAKE_C_COMPILER=$GCC_COMMAND -D CMAKE_CXX_COMPILER=$GXX_COMMAND -D CMAKE_BUILD_TYPE=Release -D USE_GPU:STR=No -D USE_OMP:STR=No ..
```

File: script/build_clang.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
```
Comment (KowerKoint) [inline script/build_clang.sh:1]: Change file mode to be executable.

**Comment by KowerKoint** [comment]
Same error to @forest1040 's, and I could resolve it by replacing
```
	set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c11")
	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
```
with
```
	set(CMAKE_C_STANDARD 11)
	set(CMAKE_CXX_STANDARD 14)
```
.
I do not know why.

Then, I fail on `pip install .`
This seems to be because `gcc` is specified in `setup.py` line 106.
You have to fix it too.

**Comment by forest1040** [comment]
@KowerKoint @ikanago 
Thank you!

> This seems to be because gcc is specified in setup.py line 106.

I changed it at local my Mac. It's build success!!

**Comment by ikanago** [comment]
I fixed `CMAKE_{C,CXX}_STANDARD` and specify `clang++`(not just `clang`) for `CXX_COMPILER` in macOS CI, then all the test passes.
I postpone organizing wheel build for another PR, so please approve and merge this PR if there's no problem.

**Comment by ikanago** [comment]
Clang complains there's unknown warning options which is available gcc.
I leave it for now and I'm going to fix them in #456.

**Review by forest1040 — APPROVED** [review]
@ikanago Thank you!
My local mac build success and all test ok!
Please tell me.

> This seems to be because gcc is specified in setup.py line 106.

Is this fix not nessesary?
I don't know which setup.py to use gcc or clang?

**Comment by KowerKoint** [comment]
`setup.py` specifies `C_COMPILER` default as `gcc`, so just running `pip install .` in M1 fails (`C_COMPILER=clang CXX_COMPILER=clang++ pip install .` works.)
I want to specify `clang` `clang++` as default values when building `arm64` wheels, but I found it hard to detect target architecture in `setup.py`.
I concluded it to be not necessary.

**Review by KowerKoint — APPROVED** [review]

**Review by kotamanegi — APPROVED** [review]
LGTM:+1:
Thank you for your hard work to clang compatibility!
However, I found 1 minor issue that needs to be addressed.

File: CMakeLists.txt [hunk 4]
```diff
@@ -30,14 +20,47 @@ if(("${CMAKE_HOST_SYSTEM_PROCESSOR}" STREQUAL "aarch64") OR
...
+	message(STATUS "MSVC_VERSION = ${MSVC_VERSION}")
+	if(${MSVC_VERSION} VERSION_LESS 1900)
+		message(FATAL_ERROR "MSVC >= 2015 is required.")
+	endif()
+elseif ((${CMAKE_CXX_COMPILER_ID} STREQUAL "Clang") OR (${CMAKE_CXX_COMPILER_ID} STREQUAL "AppleClang"))
+	set(DEFAULT_USE_OMP No)
+	if(${COVERAGE})
+		message(FATAL_ERROR "Coverage is not supported for clang. Do not set COVERAGE=Yes on this environment.")
+	endif()
+elseif (${CMAKE_CXX_COMPILER_ID} STREQUAL "Intel")
```
[the remark this answers is not in the document]
Comment (ikanago) [reply CMakeLists.txt]: Removed `elseif` clause on line 57 which checks Intel compiler.

**Pull request merged**